shared/app.ts

```diff
@@ -0,0 +1,196 @@
+    adhocPlaybackMode?: AdhocChordPlaybackMode;
+
+    private activeMode: ApplicationModeManager<AdhocProgressionState> = this.adhocCompositionMode!;
+
+    chordSupervisor = new OutputChordSupervisor(this.midiService);
+
+    deps = {
+        midi: this.midi,
+        stdin: this.stdin,
+        config: this.config,
+        userData: this.userData,
+    };
+
+    services = {
+        midi: this.midiService,
+        userData: this.userData,
+    };
+
+    getUserData = () => this.userData;
+    getConfig = () => this.config;
+    getProgressionState: () => ProgressionState | undefined = () => this.progressionMode?.getState();
+    getAdhocState: () => AdhocProgressionState | undefined = () => {
+        if (this.adhocCompositionMode) {
+            this.adhocCompositionMode.getState();
+        }
+
+        if (this.adhocPlaybackMode) {
+            return this.adhocPlaybackMode.getState();
+        }
+
+        return undefined;
+    };
+
+    getState = (): GlobalState => {
+        return {
+            config: this.config,
+            userData: this.userData,
+            // progression: this.getProgressionState(),
+            adhocState: this.getAdhocState(),
+        };
+    };
+
+    broadcastState = (_flash?: boolean) => {
+        const state = this.getState();
+        this.globalStateSubject.next(state);
+    };
+
+    subscribeToGlobalState = (callback: (subjectMessage: GlobalState) => void) => {
+        return this.globalStateSubject.subscribe(callback);
+    };
+
+    changeModeAdhocPlayback = (state: AdhocProgressionState) => {
+        if (this.adhocPlaybackMode) {
+            return;
+        }
+
+        this.adhocCompositionMode?.close();
+        this.adhocCompositionMode = undefined;
+
+        const newState: AdhocProgressionState = {...state, mode: 'playback'};
+        this.adhocPlaybackMode = new AdhocChordPlaybackMode(newState, this.midiService, this);
+        this.activeMode = this.adhocPlaybackMode;
+
+        this.broadcastState();
+    };
+
+    changeModeAdhocComposition = () => {
+        this.adhocPlaybackMode?.close();
+        this.adhocPlaybackMode = undefined;
+        this.adhocCompositionMode?.close();
+        this.adhocCompositionMode = undefined;
+
+        this.midiService.notesOffAll();
+
+        this.adhocCompositionMode = new AdhocChordCompositionMode(this.midiService, this);
+        this.activeMode = this.adhocCompositionMode;
+
+        this.broadcastState();
+    };
+
+    actions = {
+        toggleDrumsColorAction: () => this.progressionMode?.toggleDrumColorAction(),
+        toggleDrumsMusicAction: () => this.progressionMode?.toggleDrumMusicAction(),
+        setRandomColor: () => this.wledService.setRandomColor(),
+        setRandomEffect: () => this.wledService.setRandomEffect(),
+        noteOffAll: () => {
+            this.midiService.notesOffAll();
+            this.chordSupervisor.notesOffAll();
+        },
```

Comment:
Gotta make sure the local mode calls notes off when the user refreshes. Or make it easy to do "note off all" somehow